src/circuit/mod.rs

```diff
@@ -1301,9 +1306,27 @@ mod test {
         (verification_key, proof)
     }
 
+    fn check_if_crs_exists_and_download(n: usize) {
+        let file_path = &format!("setup_2^{}.key", n);
+        let file_exists = std::path::Path::new(file_path).exists();
+    
+        if !file_exists {
+            println!("CRS not found  .. downloading CRS from https://universal-setup.ams3.digitaloceanspaces.com/ ");
+            let download_link = &format!("https://universal-setup.ams3.digitaloceanspaces.com/setup_2^{}.key", n);
+            let crs_file_destination = &format!("src/circuit/setup2^{}.key",n);
+            let mut resp = reqwest::get(download_link).expect("Unable to download CRS");
+            let mut out = File::create(crs_file_destination).expect("failed to create file");
+            io::copy(&mut resp, &mut out).expect("failed to copy content");
+        } else {
+            println!("CRS found .. moving on!");
+        }
+    }
+
     fn open_crs_for_log2_of_size(n: usize) -> Crs::<Bn256, CrsForMonomialForm> {
-        let base_path = std::path::Path::new("/Users/alexvlasov/Downloads/setup/processed");
-        let full_path = base_path.join(&format!("setup_2^{}.key", n));
+        // let base_path = std::path::Path::new("/Users/alexvlasov/Downloads/setup/processed");
+        check_if_crs_exists_and_download(n);
+        let crs_file_name = &format!("setup_2^{}.key", n);
+        let full_path = std::path::Path::new(crs_file_name);
```

Comment:
I would compress this to one line, and make the function return the full path. See comment above

`let full_path = check_if_crs_exists_and_download(n)`

Reply:
`check_if_crs_exists_and_download` return `doesnt` return a value so i would need to make it return a `Option<T>` ?

Reply:
I meant that you could change the function signature so that it outputs the path (whether the file existed or not)